### PR TITLE
Initial implementation of filter concept

### PR DESCRIPTION
**Pull request by DavidMinarsch**
## Proposed changes

This PR introduces the concept of the `Filter`. The filter is responsible for keeping track of inactive skill components and for routing internal messages.

## Fixes

It addresses #295 

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../master/CONTRIBUTING.rst) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that the documentation about the `aea cli` tool works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

-

### CODE REVIEW

File: aea/decision_maker/base.py [hunk 1]
```diff
@@ -41,6 +41,7 @@
 ExchangeParams = Dict[str, float]   # a map from identifier to quantity
 
 QUANTITY_SHIFT = 100
+INTERNAL_PROTOCOL_ID = 'internal'
```
Comment (DavidMinarsch) [inline aea/decision_maker/base.py]: Not sure where to place this constant yet, hence duplicated throughout.

File: aea/aea.py [hunk 2]
```diff
@@ -160,33 +167,25 @@ def handle(self, envelope: Envelope) -> None:
             error_handler.send_invalid_message(envelope)    # pragma: no cover
             return                                          # pragma: no cover
 
-        handlers = self.resources.handler_registry.fetch(protocol.id)
+        handlers = self.filter.get_active_handlers(protocol.id)
         if handlers is None:
             if error_handler is not None:
                 error_handler.send_unsupported_skill(envelope)
```
Comment (marcofavorito) [inline aea/aea.py]: should it really be an "unsupported skill" error? Should discern between "inactive" handler and "unsupported" handler?

File: aea/decision_maker/base.py [hunk 3]
```diff
@@ -299,11 +300,9 @@ def execute(self) -> None:
...
         :return: None
         """
-        counter = 0
-        while not self.message_in_queue.empty() and counter < self._max_reactions:
-            counter += 1
+        while not self.message_in_queue.empty():
             message = self.message_in_queue.get_nowait()  # type: Optional[Message]
-            if message is not None:
+            if message is not None and message.protocol_id == INTERNAL_PROTOCOL_ID:
                 self.handle(message)
```
Comment (marcofavorito) [inline aea/decision_maker/base.py]: As a suggestion, I'd put the following code
```python
else:
    logger.warning("Message received by the decision maker is not labelled as 'internal'")
```

or something like that. Might be useful to have for debugging.

File: aea/protocols/base.py [hunk 4]
```diff
@@ -33,6 +33,8 @@
 class Message:
     """This class implements a message."""
 
+    protocol_id = 'base'
```
Comment (marcofavorito) [inline aea/protocols/base.py]: does that suggest we could also support the serialization for the `base` protocol? 
E.g. we could recommend using `JSONSerializer` to serialize objects of type `Message`.

File: aea/aea.py (hunk 2 above)
Comment (DavidMinarsch) [reply aea/aea.py]: I am not sure. Basically, this is just telling the other agent: "I cannot deal with this". We probably should keep private to the agent the info why that is (i.e. unsupported or not active skill). So the point I would make is that it should maybe be renamed.

File: aea/protocols/base.py (hunk 4 above)
Comment (DavidMinarsch) [reply aea/protocols/base.py]: No, we don't support that at the moment. But we need to provide a class variable in the base class for typing consistency.

File: aea/registries/base.py [hunk 5]
```diff
@@ -533,3 +546,76 @@ def teardown(self):
...
+    def __init__(self, resources: Resources, decision_maker_out_queue: Queue):
+        """
+        Instantiate the filter.
+
+        :param resources: the resources
+        :param decision_maker_out_queue: the decision maker queue
+        """
+        self._resources = resources
+        self._decision_maker_out_queue = decision_maker_out_queue
+        # TODO: self._inactive_handlers = {}  # type: Dict[SkillId, List[HandlerId]]
```
Comment (DavidMinarsch) [inline aea/registries/base.py]: Note, @marcofavorito how this demonstrates the need for `HandlerId`, `BehaviourId` and `TaskId`

**Pull request merged**